### PR TITLE
Virtual association

### PR DESCRIPTION
**Pull request by dakota**
Fixes an issue where an entity has a getter method for a association (For use with lazy loading for example).

With the recent virtual property change, the annotator would create duplicate annotations.

### CODE REVIEW

File: src/Annotator/EntityAnnotator.php [hunk 1]
```diff
@@ -57,15 +57,11 @@ public function annotate($path) {
 		$propertyHintMap = $helper->buildEntityPropertyHintTypeMap($schema);
 		$propertyHintMap = $this->buildExtendedEntityPropertyHintTypeMap($schema, $helper) + $propertyHintMap;
 		$propertyHintMap += $this->buildVirtualPropertyHintTypeMap($content);
+		$propertyHintMap += $helper->buildEntityAssociationHintTypeMap($schema);
```
Comment (dereuromark) [inline src/Annotator/EntityAnnotator.php]: Didnt  $helper->propertyHints($associationHintMap)) get lost now?
Please also keep the noise changes at a minimum for the surrounding code. "composer cs-fix" helps.

Comment (dereuromark) [reply src/Annotator/EntityAnnotator.php]: Looking at the merge I would actually prefer the association ones before the virtual ones.
We need to change the merging op.

Comment (dakota) [reply src/Annotator/EntityAnnotator.php]: Yup, the association hints are added to the `$propertyHintMap` instead before building the hints (Line 64).

Sorry about the noise, I forgot you use a different standard than everybody else :)

File: src/Annotator/ModelAnnotator.php [hunk 2]
```diff
@@ -200,7 +204,7 @@ protected function _getAssociations(AssociationCollection $tableAssociations) {
 			$association = $tableAssociations->get($key);
 			$type = get_class($association);
 
-			$name = $association->getAlias();
+			list(, $name) = pluginSplit($association->getAlias());
```
Comment (dereuromark) [inline src/Annotator/ModelAnnotator.php]: An alias can have a dot prefix?

File: src/Annotator/ModelAnnotator.php [hunk 3]
```diff
@@ -257,8 +261,10 @@ protected function throughAlias(BelongsToMany $association) {
...
 		$tablesNames = array_map('Cake\Utility\Inflector::underscore', [
-			$association->getSource()->getTable(),
-			$association->getTarget()->getTable()
+			$association->getSource()
+				->getTable(),
+			$association->getTarget()
+				->getTable()
 		]);
 
 		sort($tablesNames);
```
Comment (dakota) [inline src/Annotator/ModelAnnotator.php]: These changes are unrelated to this PR (Will be creating a different PR for these). Will fix it up, sorry about that.

**Comment by dereuromark** [comment]
You can force push a cleaned up, no probs.

File: src/Annotator/EntityAnnotator.php (hunk 1 above)
Comment (dereuromark) [reply src/Annotator/EntityAnnotator.php]: $propertyHintMap += $this->buildVirtualPropertyHintTypeMap()
$propertyHintMap += $helper->buildEntityAssociationHintTypeMap()

So virtual before assoc, fair enough then :+1:

File: src/Annotator/EntityAnnotator.php [hunk 4]
```diff
@@ -57,15 +57,11 @@ public function annotate($path) {
...
 
 		$propertyHintMap = array_filter($propertyHintMap);
 
 		$annotations = $helper->propertyHints($propertyHintMap);
-		$associationHintMap = $helper->buildEntityAssociationHintTypeMap($schema);
-
-		if ($associationHintMap) {
-			$annotations = array_merge($annotations, $helper->propertyHints($associationHintMap));
```
Comment (dereuromark) [inline src/Annotator/EntityAnnotator.php]: array_merge() often kills you in the end .. :P
I should have guessed it here, too

**Comment by dakota** [comment]
@dereuromark All is fixed now :)
